sequence-generator/src/main/java/org/terracotta/management/sequence/Defaults.java

```diff
@@ -0,0 +1,163 @@
+
+  static final int INSTANCE_BITLENGTH = 18;
+  static final int INSTANCE_BITMASK = (1 << INSTANCE_BITLENGTH) - 1;
+
+  static final NodeIdSource MAC_PID_NODE_ID_SOURCE = new NodeIdSource() {
+    @Override
+    public long getNodeId() {
+      // at least 6 bytes
+      byte[] mac = readFirstNonLoopbackMacAddress();
+      long nodeId = 0;
+      // we consider the 6 last bytes only,
+      for (int i = Math.max(0, mac.length - 6); i < mac.length; i++) {
+        nodeId = (nodeId << 8) | (mac[i] & 0XFF);
+      }
+      // keeps a positive node id with the 0x7fffffffffffffffL mask
+      return ((nodeId << PID_BITLENGTH) & Long.MAX_VALUE) | (readPID() & PID_BITMASK);
+    }
+  };
+
+  static final NodeIdSource BEST_NODE_ID_SOURCE = new NodeIdSource() {
+    final NodeIdSource delegate = findBest(NodeIdSource.class, MAC_PID_NODE_ID_SOURCE);
+
+    @Override
+    public long getNodeId() {
+      return delegate.getNodeId();
+    }
+  };
+
+  static final TimeSource SYSTEM_TIME_SOURCE = new TimeSource() {
+    @Override
+    public long getTimestamp() {
+      return System.currentTimeMillis();
+    }
+  };
+
+  static final TimeSource BEST_TIME_SOURCE = new TimeSource() {
+    final TimeSource delegate = findBest(TimeSource.class, SYSTEM_TIME_SOURCE);
+
+    @Override
+    public long getTimestamp() {
+      return delegate.getTimestamp();
+    }
+  };
+
+  private static final Comparator<NetworkInterface> NETWORK_INTERFACE_COMPARATOR = new Comparator<NetworkInterface>() {
+    @Override
+    public int compare(NetworkInterface o1, NetworkInterface o2) {
+      return o1.getName().compareTo(o2.getName());
+    }
+  };
+
+  static byte[] readFirstNonLoopbackMacAddress() {
+    List<NetworkInterface> networkInterfaces;
+    try {
+      networkInterfaces = Collections.list(NetworkInterface.getNetworkInterfaces());
+    } catch (SocketException e) {
+      throw new IllegalStateException("Machine has no network interfaces!", e);
+    }
+
+    // order interfaces by name
+    Collections.sort(networkInterfaces, NETWORK_INTERFACE_COMPARATOR);
+
+    // find the first accessible non-loopback interface having a mac address
+    // we try first to skip all virtual interfaces since they can be easily dynamically created on-demand
+    // the goal of this method is to try return the same value each time as possible
+    // we consider only "parent" interfaces
+    for (NetworkInterface networkInterface : networkInterfaces) {
+      try {
+        byte[] mac;
+        if (!networkInterface.isLoopback()
+            && !networkInterface.isPointToPoint()
+            && !networkInterface.isVirtual()
+            && networkInterface.getParent() == null
+            && (mac = networkInterface.getHardwareAddress()) != null
+            && mac.length >= 6) {
+          return mac;
+        }
+      } catch (SocketException ignored) {
+      }
+    }
+
+    // if we do not succeed, we enlarge our search
```

Comment:
yes. actually what I wanted to is first find a REAL interface (i.e. eth0, eth1, wlan0, etc). If i cannot, then I'll just take the first non-loopback interface I find.

The goal is just to be able to get a mac address, and, if we can, _return always the same._

So the ordering and filtering tries to always return the same first. Virtual interfaces can be added and removed dynamically. So I skip them first. Anyway, virtual interfaces have parents. So it means that if there are some virtual interfaces, there should logically be some existing real interfaces (i.e. eth0).
So this second loop is just in case the first one is too restrictive, then the second loop does not filter at all. In most cases, the second loop will be skipped.